GVRf/Framework/jni/engine/renderer/renderer.cpp

```diff
@@ -51,25 +51,108 @@ void Renderer::renderCamera(std::shared_ptr<Scene> scene,
+
         std::vector < std::shared_ptr < SceneObject >> scene_objects =
                 scene->getWholeSceneObjects();
         std::vector < std::shared_ptr < RenderData >> render_data_vector;
-        for (auto it = scene_objects.begin(); it != scene_objects.end(); ++it) {
+
+        for (auto it = scene_objects.begin(); it != scene_objects.end(); ++it)
+        {
             std::shared_ptr<RenderData> render_data = (*it)->render_data();
-            if (render_data != 0) {
-                if (render_data->material() != 0) {
-                    render_data_vector.push_back(render_data);
+            if (render_data != 0)
+            {
+                if (render_data->material() != 0)
+                {
+                    // Check for frustum culling flag
+                    if(scene->get_frustum_culling())
+                    {
+                        // Frustum culling setup
+                        std::shared_ptr<Mesh> currentMesh = render_data->mesh();
+                        std::vector<float> bounding_box_info = currentMesh->getBoundingBoxInfo();
+                        glm::mat4 model_matrix_tmp(render_data->owner_object()->transform()->getModelMatrix());
+                        glm::mat4 mvp_matrix_tmp( vp_matrix * model_matrix_tmp);
+
+                        // Frustum
+                        float frustum[6][4];
+
+                        // Matrix to array
+                        float mvp_matrix_array[16] = { 0.0 };
+
+                        const float *mat_to_array = (const float*) glm::value_ptr(mvp_matrix_tmp);
+                        memcpy(mvp_matrix_array, mat_to_array, sizeof(float)*16);
+
+                        // Build the frustum
+                        build_frustum(frustum, mvp_matrix_array);
+                        // Check for being inside or outside frustum
+                        bool is_inside = is_cube_in_frustum(frustum, bounding_box_info);
+                        // Only push those scene objects that are inside of the frustum
+                        if(is_inside)
+                        {
+                            (*it)->set_in_frustum();
+                            bool visible = (*it)->visible();
+
+                            //If visibility flag was set by an earlier occlusion query,
+                            //turn visibility on for the object
+                            if(visible)
+                                render_data_vector.push_back(render_data);
+#if _GVRF_USE_GLES3_
+                            if (render_data->material() != 0)
+                            {
+                                if(scene->get_occlusion_culling())
+                                {
+                                    //Setup basic bounding box and material
+                                    std::shared_ptr<RenderData> bounding_box_render_data(new RenderData());
+                                    std::shared_ptr<Mesh> bounding_box_mesh = render_data->mesh()->getBoundingBox();
```

Comment:
is this the only time getBoundingBox() is called?  it seems like calling it is necessary for the bounding box info in the frustum cull portion of the code.  i might be blind as to where else it is called though.